### PR TITLE
refactor!: simplify namespaces to make public API more pythonic

### PR DESCRIPTION
**Pull request by federicobond**
For reviewing, it's best to follow the changes commit by commit.

This reduces the number of public namespaces in the sdk, following what we agreed on https://github.com/open-feature/python-sdk/issues/101

I consolidated some modules into a single file when it made sense. The alternative would have been to prefix the internal namespace files with underscores and re-export the relevant symbols from `__init__.py` but it added a lot of boilerplate with little benefit. If the modules grow in size to make this necessary, it can be done later without breaking compatibility with existing users.

### CODE REVIEW

**Review by beeme1mr — APPROVED** [review]
This looks great, thank you! Please make sure to update the examples on the readme.

**Comment by toddbaert** [comment]
I'll review this first thing next week. Thanks @federicobond

File: tests/provider/test_in_memory_provider.py [hunk 1]
```diff
@@ -1,8 +1,7 @@
 from numbers import Number
 
-from open_feature.exception.error_code import ErrorCode
-from open_feature.flag_evaluation.reason import Reason
-from open_feature.flag_evaluation.resolution_details import FlagResolutionDetails
+from open_feature.exception import ErrorCode
+from open_feature.flag_evaluation import FlagResolutionDetails, Reason
```
Comment (toddbaert) [inline tests/provider/test_in_memory_provider.py:4]: Ya these all feel A LOT better.

**Comment by toddbaert** [comment]
> If the modules grow in size to make this necessary, it can be done later without breaking compatibility with existing users.

:+1:

**Review by toddbaert — APPROVED** [review]
LGTM!

**Comment by toddbaert** [comment]
Obviously this is breaking, but I have no issue merging it. I'll do so in the next few days unless I hear objections from the other (more Python-literate) reviewers.

**Pull request merged**